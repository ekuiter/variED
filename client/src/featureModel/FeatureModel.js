import React from 'react';
import VerticalTreeLayout from './treeLayout/VerticalTreeLayout';
import HorizontalTreeLayout from './treeLayout/HorizontalTreeLayout';

class FeatureModel extends React.Component {
    static defaultProps = {layout: 'verticalTree'};

    static layoutMap = {
        verticalTree: VerticalTreeLayout,
        horizontalTree: HorizontalTreeLayout
    };

    layoutRef = React.createRef();

    key(props) {
        // The key uniquely identifies the layout component instance. If the key changes,
        // the instance is unmounted and a new one is mounted. This is useful for forcing
        return props.debug; // rerenders, i.e. when the debug flag changes.
    }

    render() {
        const {layout, ...props} = this.props,
            LayoutComponent = FeatureModel.layoutMap[layout];

        return (
            <LayoutComponent key={this.key(props)} ref={this.layoutRef} {...props} />
        );
    }

    export() {
        const layout = this.layoutRef.current;
        if (layout && layout.canExport())
            return layout.export().then(console.log);
        else
            return Promise.reject('can not export feature model');
    }
}

export default FeatureModel;