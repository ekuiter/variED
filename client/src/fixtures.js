
export const invalidFeatureModel1 = {};

export const invalidFeatureModel2 = {
    'struct': [],
    'constraints': [],
    'properties': [],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};

export const validFeatureModel = {
    'struct': [
        {
            'name': 'Eclipse',
            'mandatory': true,
            'description': 'his is a feature which has a description text',
            'type': 'and',
            'children': [
                {'type': 'feature', 'name': 'JDT'},
                {'type': 'feature', 'name': 'CDT'},
                {'type': 'feature', 'name': 'AJDT'},
                {
                    'name': 'FeatureModeling',
                    'type': 'and',
                    'children': [
                        {'type': 'feature', 'name': 'CIDE'},
                        {'type': 'feature', 'name': 'FAMILIAR'},
                        {
                            'name': 'FeatureIDE',
                            'type': 'or',
                            'children': [
                                {'type': 'feature', 'name': 'AHEAD'},
                                {'type': 'feature', 'name': 'FeatureHouse'},
                                {'type': 'feature', 'name': 'FeatureCpp'},
                                {'type': 'feature', 'name': 'DeltaJ'},
                                {'type': 'feature', 'name': 'AspectJ'},
                                {'type': 'feature', 'name': 'Munge'},
                                {'type': 'feature', 'name': 'Antenna'},
                                {'type': 'feature', 'name': 'DeltaMontiArc'}
                            ]
                        },
                        {'type': 'feature', 'name': 'ExtendedFM'},
                        {'type': 'feature', 'name': 'MoSoPoLiTe'}
                    ]
                },
                {'type': 'feature', 'name': 'DeltaJEclipsePlugin'}
            ]
        }
    ],
    'constraints': [
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {
                            'type': 'disj',
                            'children': [
                                {'type': 'var', 'var': 'AHEAD'},
                                {
                                    'type': 'disj',
                                    'children': [
                                        {'type': 'var', 'var': 'FeatureHouse'},
                                        {
                                            'type': 'disj',
                                            'children': [
                                                {'type': 'var', 'var': 'Munge'},
                                                {'type': 'var', 'var': 'Antenna'}
                                            ]
                                        }
                                    ]
                                }
                            ]
                        },
                        {'type': 'var', 'var': 'JDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'FeatureCpp'},
                        {'type': 'var', 'var': 'CDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'AspectJ'},
                        {'type': 'var', 'var': 'AJDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'DeltaJ'},
                        {'type': 'var', 'var': 'DeltaJEclipsePlugin'}
                    ]
                }
            ]
        }
    ],
    'properties': [],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};