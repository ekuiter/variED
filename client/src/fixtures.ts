import {SerializedFeatureModel} from './modeling/types';

/**
 * Some fixtures for use in tests.
 */

export const invalidFeatureModel1 = <SerializedFeatureModel>{};

export const invalidFeatureModel2 = <any>{
    'struct': [],
    'constraints': [],
    'properties': [],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};

export const validFeatureModel = <SerializedFeatureModel>{
    'struct': [
        {
            'ID': 'Eclipse',
            'name': 'Eclipse',
            'mandatory': true,
            'description': 'this is a feature which has a description text',
            'type': 'and',
            'children': [
                {'type': 'feature', 'ID': 'JDT', 'name': 'JDT'},
                {'type': 'feature', 'ID': 'CDT', 'name': 'CDT'},
                {'type': 'feature', 'ID': 'AJDT', 'name': 'AJDT'},
                {
                    'ID': 'FeatureModeling',
                    'name': 'FeatureModeling',
                    'type': 'and',
                    'children': [
                        {'type': 'feature', 'ID': 'CIDE', 'name': 'CIDE'},
                        {'type': 'feature', 'ID': 'FAMILIAR', 'name': 'FAMILIAR'},
                        {
                            'ID': 'FeatureIDE',
                            'name': 'FeatureIDE',
                            'description': 'A sample description',
                            'type': 'or',
                            'children': [
                                {'type': 'feature', 'ID': 'AHEAD', 'name': 'AHEAD'},
                                {'type': 'feature', 'ID': 'FeatureHouse', 'name': 'FeatureHouse'},
                                {'type': 'feature', 'ID': 'FeatureCpp', 'name': 'FeatureCpp'},
                                {'type': 'feature', 'ID': 'DeltaJ', 'name': 'DeltaJ'},
                                {'type': 'feature', 'ID': 'AspectJ', 'name': 'AspectJ'},
                                {'type': 'feature', 'ID': 'Munge', 'name': 'Munge'},
                                {'type': 'feature', 'ID': 'Antenna', 'name': 'Antenna'},
                                {'type': 'feature', 'ID': 'DeltaMontiArc', 'name': 'DeltaMontiArc'}
                            ]
                        },
                        {'type': 'feature', 'ID': 'ExtendedFM', 'name': 'ExtendedFM'},
                        {'type': 'feature', 'ID': 'MoSoPoLiTe', 'name': 'MoSoPoLiTe'}
                    ]
                },
                {'type': 'feature', 'ID': 'DeltaJEclipsePlugin', 'name': 'DeltaJEclipsePlugin'}
            ]
        }
    ],
    'constraints': [
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {
                            'type': 'disj',
                            'children': [
                                {'type': 'var', 'var': 'AHEAD'},
                                {
                                    'type': 'disj',
                                    'children': [
                                        {'type': 'var', 'var': 'FeatureHouse'},
                                        {
                                            'type': 'disj',
                                            'children': [
                                                {'type': 'var', 'var': 'Munge'},
                                                {'type': 'var', 'var': 'Antenna'}
                                            ]
                                        }
                                    ]
                                }
                            ]
                        },
                        {'type': 'var', 'var': 'JDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'FeatureCpp'},
                        {'type': 'var', 'var': 'CDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'AspectJ'},
                        {'type': 'var', 'var': 'AJDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'DeltaJ'},
                        {'type': 'var', 'var': 'DeltaJEclipsePlugin'}
                    ]
                }
            ]
        }
    ],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};

export const validFeatureModelWithRemovedFeatures = <SerializedFeatureModel>{
    'struct': [
        {
            'name': 'Eclipse',
            'mandatory': true,
            'description': 'this is a feature which has a description text',
            'type': 'and',
            'children': [
                {'type': 'feature', 'name': 'JDT'},
                {'type': 'feature', 'name': 'CDT'},
                {'type': 'feature', 'name': 'AJDT'},
                {
                    'name': 'FeatureModeling',
                    'type': 'and',
                    'children': [
                        {'type': 'feature', 'name': 'CIDE'},
                        {'type': 'feature', 'name': 'FAMILIAR'},
                        {'type': 'feature', 'name': 'AHEAD'},
                        {'type': 'feature', 'name': 'FeatureHouse'},
                        {'type': 'feature', 'name': 'FeatureCpp'},
                        {'type': 'feature', 'name': 'AspectJ'},
                        {'type': 'feature', 'name': 'Munge'},
                        {'type': 'feature', 'name': 'Antenna'},
                        {'type': 'feature', 'name': 'DeltaMontiArc'},
                        {'type': 'feature', 'name': 'ExtendedFM'},
                        {'type': 'feature', 'name': 'MoSoPoLiTe'}
                    ]
                }
            ]
        }
    ],
    'constraints': [
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {
                            'type': 'disj',
                            'children': [
                                {'type': 'var', 'var': 'AHEAD'},
                                {
                                    'type': 'disj',
                                    'children': [
                                        {'type': 'var', 'var': 'FeatureHouse'},
                                        {
                                            'type': 'disj',
                                            'children': [
                                                {'type': 'var', 'var': 'Munge'},
                                                {'type': 'var', 'var': 'Antenna'}
                                            ]
                                        }
                                    ]
                                }
                            ]
                        },
                        {'type': 'var', 'var': 'JDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'FeatureCpp'},
                        {'type': 'var', 'var': 'CDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'AspectJ'},
                        {'type': 'var', 'var': 'AJDT'}
                    ]
                }
            ]
        }
    ],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};