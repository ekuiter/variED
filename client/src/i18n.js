import objectPath from 'object-path';
import React from 'react';
import constants from './constants';
import {Link} from 'office-ui-fabric-react/lib/Link';

export const strings = {
    panels: {
        aboutPanel: {
            title: 'About',
            content: (
                <div>
                    <h3>variED: The variability editor</h3>
                    <p>View, edit and analyze feature models in the browser - with support for real-time
                        collaboration.</p>
                    <p>This project is released under the <Link href={constants.panels.aboutPanel.licenseUri}
                                                                target="_blank">LGPL v3 license</Link>.</p>
                    <p><Link href={constants.panels.aboutPanel.githubUri} target="_blank">View source code on
                        GitHub</Link></p>
                </div>
            )
        },
        settingsPanel: {
            title: 'Settings',
            toggleOn: 'On',
            toggleOff: 'Off',
            apply: 'Apply',
            customizeColors: 'Customize colors',
            resetToDefaults: 'Reset to defaults',
            headings: {
                featureDiagram: 'Feature diagram',
                features: 'Features',
                edges: 'Edges',
                verticalTree: 'Vertical tree',
                horizontalTree: 'Horizontal tree'
            },
            labels: {
                featureDiagram: {
                    font: {
                        family: 'Font family',
                        size: 'Font size'
                    },
                    treeLayout: {
                        debug: 'Show developer information',
                        useTransitions: 'Animate feature model changes',
                        node: {
                            paddingX: 'Horizontal padding',
                            paddingY: 'Vertical padding',
                            strokeWidth: 'Border thickness',
                            abstractFill: 'Abstract feature (fill)',
                            abstractStroke: 'Abstract feature (border)',
                            concreteFill: 'Concrete feature (fill)',
                            concreteStroke: 'Concrete feature (border)',
                            visibleFill: 'Visible feature (text)',
                            hiddenFill: 'Hidden feature (text)'
                        },
                        link: {
                            circleRadius: 'Circle radius',
                            stroke: 'Edge',
                            strokeWidth: 'Thickness'
                        },
                        vertical: {
                            marginX: 'Horizontal gap',
                            layerHeight: 'Vertical gap',
                            groupRadius: 'Group radius'
                        },
                        horizontal: {
                            marginY: 'Vertical gap',
                            layerMargin: 'Horizontal gap'
                        },
                    }
                }
            },
            errors: {
                fontNotInstalled: 'This font is not installed on your system.'
            }
        },
        featurePanel: {
            title: 'Feature',
            edit: 'Edit',
            noDescriptionSet: 'No description set.'
        }
    },
    dialogs: {
        featureRenameDialog: {
            title: 'Rename feature',
            rename: 'Rename'
        },
        featureSetDescriptionDialog: {
            title: 'Set feature description',
            rename: 'Save'
        }
    },
    featureDiagram: {
        commands: {
            undo: 'Undo',
            redo: 'Redo',
            setLayout: 'Layout',
            selection: (isSelectMultipleFeatures, selectedFeatureNames) =>
                isSelectMultipleFeatures
                    ? `Feature selection (${selectedFeatureNames.length})`
                    : 'Select multiple features',
            verticalTree: 'Vertical tree',
            horizontalTree: 'Horizontal tree',
            feature: {
                new: 'New',
                newFeatureBelow: 'New feature below',
                remove: 'Remove',
                details: 'Details',
                rename: 'Rename',
                setDescription: 'Set description',
                properties: 'Properties',
                abstract: 'Abstract',
                concrete: 'Concrete',
                hidden: 'Hidden',
                mandatory: 'Mandatory',
                optional: 'Optional',
                and: 'And',
                or: 'Or',
                alternative: 'Alternative'
            },
            features: {
                selectAll: 'Select all',
                deselectAll: 'Deselect all',
                newFeatureAbove: 'New feature above'
            }
        }
    }
};

export default {
    t(...paths) {
        const path = paths.join('.');
        if (!objectPath.has(strings, path))
            throw new Error(`string ${path} does not exist`);
        return objectPath.get(strings, path);
    }
};