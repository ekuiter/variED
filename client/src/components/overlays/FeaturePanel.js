import React from 'react';
import {Panel, PanelType} from 'office-ui-fabric-react/lib/Panel';
import i18n from '../../i18n';
import {CommandBar} from 'office-ui-fabric-react/lib/CommandBar';
import contextualMenuItems from '../contextualMenuItems';
import PropTypes from 'prop-types';
import {FeatureModelType} from '../../server/FeatureModel';
import FeatureComponent from './FeatureComponent';

const buttonStyles = {root: {backgroundColor: 'transparent'}},
    transparentItems = items => items;

class FeaturePanel extends FeatureComponent({onDismissProp: 'onDismissed'}) {
    onRenderFooterContent = () => (
        <CommandBar
            items={transparentItems([
                contextualMenuItems.featureDiagram.feature.new(this.props.featureName, this.props.onDismissed),
                contextualMenuItems.featureDiagram.feature.remove(this.props.featureName, this.props.onDismissed)
            ])}
            overflowItems={[
                contextualMenuItems.featureDiagram.feature.rename(this.props.featureName, this.props.onShowOverlay),
                contextualMenuItems.featureDiagram.feature.setDescription(this.props.featureName, this.props.onShowOverlay),
                contextualMenuItems.featureDiagram.feature.properties(this.feature)
            ]}
            overflowButtonProps={{styles: buttonStyles}}
            styles={{root: {margin: '0 -40px', padding: '0 35px'}}}/>
    );

    renderIfFeature(feature) {
        return (
            <Panel
                isOpen={this.props.isOpen}
                type={PanelType.smallFixedFar}
                onDismissed={this.props.onDismissed}
                isLightDismiss={true}
                headerText={
                    <span>
                        {i18n.t('panels.featurePanel.title')}: <strong>{feature.name}</strong>
                    </span>}
                onRenderFooterContent={this.onRenderFooterContent}>
                <p>{feature.description || <em>{i18n.t('panels.featurePanel.noDescriptionSet')}</em>}</p>
            </Panel>
        );
    }
}

FeaturePanel.propTypes = {
    onDismissed: PropTypes.func.isRequired,
    featureModel: FeatureModelType.isRequired,
    featureName: PropTypes.string.isRequired,
    isOpen: PropTypes.bool.isRequired,
    onShowOverlay: PropTypes.func.isRequired
};

export default FeaturePanel;