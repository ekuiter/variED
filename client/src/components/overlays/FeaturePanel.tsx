/**
 * A Fabric panel that shows details for a feature and available operations.
 */

import React from 'react';
import {Panel, PanelType} from 'office-ui-fabric-react/lib/Panel';
import i18n from '../../i18n';
import {CommandBar, ICommandBarItemProps} from 'office-ui-fabric-react/lib/CommandBar';
import commands from '../commands';
import FeatureComponent, {FeatureComponentProps} from './FeatureComponent';
import {Feature} from '../../modeling/types';
import {OnShowOverlayFunction, OnCollapseFeaturesFunction, OnCollapseFeaturesBelowFunction, OnExpandFeaturesFunction, OnExpandFeaturesBelowFunction, OnRemoveFeaturesFunction, OnAddFeatureBelowFunction, OnAddFeatureAboveFunction, OnSetFeatureAbstractFunction, OnSetFeatureHiddenFunction, OnSetFeatureMandatoryFunction, OnSetFeatureAndFunction, OnSetFeatureOrFunction, OnSetFeatureAlternativeFunction, OnRemoveFeaturesBelowFunction} from '../../store/types';

type Props = FeatureComponentProps & {
    onDismissed: () => void,
    isOpen: boolean,
    onShowOverlay: OnShowOverlayFunction,
    onCollapseFeatures: OnCollapseFeaturesFunction,
    onCollapseFeaturesBelow: OnCollapseFeaturesBelowFunction,
    onExpandFeatures: OnExpandFeaturesFunction,
    onExpandFeaturesBelow: OnExpandFeaturesBelowFunction,
    onRemoveFeatures: OnRemoveFeaturesFunction,
    onRemoveFeaturesBelow: OnRemoveFeaturesBelowFunction,
    onAddFeatureBelow: OnAddFeatureBelowFunction,
    onAddFeatureAbove: OnAddFeatureAboveFunction,
    onSetFeatureAbstract: OnSetFeatureAbstractFunction,
    onSetFeatureHidden: OnSetFeatureHiddenFunction,
    onSetFeatureMandatory: OnSetFeatureMandatoryFunction,
    onSetFeatureAnd: OnSetFeatureAndFunction,
    onSetFeatureOr: OnSetFeatureOrFunction,
    onSetFeatureAlternative: OnSetFeatureAlternativeFunction
};

const buttonStyles = {root: {backgroundColor: 'transparent'}},
    transparentItems = (items: ICommandBarItemProps[]) => items;

export default class extends FeatureComponent()<Props> {
    onRenderFooterContent = () => (
        <CommandBar
            items={transparentItems([
                commands.featureDiagram.feature.newMenu(this.props.featureUUID!, this.props.onAddFeatureBelow, this.props.onAddFeatureAbove, this.props.onDismissed, true),
                commands.featureDiagram.feature.removeMenu([this.feature], this.props.onRemoveFeatures, this.props.onRemoveFeaturesBelow, this.props.onDismissed, true)
            ])}
            overflowItems={[
                commands.featureDiagram.feature.collapseMenu(
                    [this.feature], this.props.onCollapseFeatures, this.props.onExpandFeatures,
                    this.props.onCollapseFeaturesBelow, this.props.onExpandFeaturesBelow, this.props.onDismissed),
                commands.featureDiagram.feature.rename(this.props.featureUUID!, this.props.onShowOverlay),
                commands.featureDiagram.feature.setDescription(this.props.featureUUID!, this.props.onShowOverlay),
                commands.featureDiagram.feature.properties([this.feature], this.props.onSetFeatureAbstract,
                    this.props.onSetFeatureHidden, this.props.onSetFeatureMandatory, this.props.onSetFeatureAnd,
                    this.props.onSetFeatureOr, this.props.onSetFeatureAlternative, this.props.onDismissed)
            ]}
            overflowButtonProps={{styles: buttonStyles}}
            styles={{root: {margin: '0 -40px', padding: '0 35px'}}}/>
    );

    renderIfFeature(feature: Feature) {
        return (
            <Panel
                isOpen={this.props.isOpen}
                type={PanelType.smallFixedFar}
                onDismissed={this.props.onDismissed}
                isLightDismiss={true}
                headerText={
                    <span>
                        {i18n.t('overlays.featurePanel.title')}: <strong>{feature.name}</strong>
                    </span> as any}
                onRenderFooterContent={this.onRenderFooterContent}>
                <p>{feature.description || <em>{i18n.t('overlays.featurePanel.noDescriptionSet')}</em>}</p>
            </Panel>
        );
    }
}