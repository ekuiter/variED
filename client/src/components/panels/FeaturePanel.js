import React from 'react';
import {Panel, PanelType} from 'office-ui-fabric-react/lib/Panel';
import i18n from '../../i18n';
import {CommandBar} from 'office-ui-fabric-react/lib/CommandBar';
import contextualMenuItems from '../contextualMenuItems';

const buttonStyles = {root: {backgroundColor: 'transparent'}},
    transparentItems = items => items;

export default class extends React.Component {
    static defaultProps = {
        isOpen: false, onDismissed: null, featureName: null, onShowDialog: null, featureModel: null
    };

    onRenderFooterContent = () => (
        <CommandBar
            items={transparentItems([
                contextualMenuItems.featureDiagram.feature.new(this.props.featureName, this.props.onDismissed),
                contextualMenuItems.featureDiagram.feature.remove(this.props.featureName, this.props.onDismissed)
            ])}
            overflowItems={[
                contextualMenuItems.featureDiagram.feature.rename(this.props.featureName, this.props.onShowDialog),
                contextualMenuItems.featureDiagram.feature.setDescription(this.props.featureName, this.props.onShowDialog)
            ]}
            overflowButtonProps={{styles: buttonStyles}}
            styles={{root: {margin: '0 -40px', padding: '0 35px'}}}/>
    );

    render() {
        const feature = this.props.featureModel && this.props.featureModel.getFeatureOrDismiss(
            this.props.featureName, this.props.isOpen, this.props.onDismissed);
        if (!feature)
            return null;
        return (
            <Panel
                isOpen={this.props.isOpen}
                type={PanelType.smallFixedFar}
                onDismissed={this.props.onDismissed}
                isLightDismiss={true}
                headerText={
                    <span>
                        {i18n.t('panels.featurePanel.title')}: <strong>{this.props.featureName}</strong>
                    </span>}
                onRenderFooterContent={this.onRenderFooterContent}>
                {feature &&
                <p>{feature.description || <em>{i18n.t('panels.featurePanel.noDescriptionSet')}</em>}</p>}
            </Panel>
        );
    }
};