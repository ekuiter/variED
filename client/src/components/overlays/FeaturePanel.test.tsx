import React from 'react';
import {shallow} from 'enzyme';
import FeaturePanel from './FeaturePanel';
import FeatureModel from '../../server/FeatureModel';
import {validFeatureModel} from '../../fixtures';
import {Panel} from 'office-ui-fabric-react/lib/Panel';
import i18n from '../../i18n';

describe('FeaturePanel', () => {
    const featurePanel = (featureName = 'FeatureIDE') => {
        const featureModel = new FeatureModel(validFeatureModel, []),
            mock = jest.fn();
        return shallow(
            <FeaturePanel
                isOpen={true}
                onDismissed={mock}
                onShowOverlay={mock}
                onCollapseFeatures={mock}
                onExpandFeatures={mock}
                onCollapseFeaturesBelow={mock}
                onExpandFeaturesBelow={mock}
                featureModel={featureModel}
                featureName={featureName}/>
        );
    };

    it('renders information for a feature with a description', () => {
        const wrapper = featurePanel(),
            headerText = shallow(wrapper.find(Panel).prop('headerText') as any);
        expect(headerText.contains('FeatureIDE')).toBe(true);
        expect(wrapper.find(Panel).contains('A sample description')).toBe(true);
        expect(wrapper).toMatchSnapshot();
    });

    it('renders information for a feature without a description', () => {
        const wrapper = featurePanel('FeatureHouse'),
            headerText = shallow(wrapper.find(Panel).prop('headerText') as any);
        expect(headerText.contains('FeatureHouse')).toBe(true);
        expect(wrapper.find(Panel).contains(i18n.t('overlays.featurePanel.noDescriptionSet'))).toBe(true);
        expect(wrapper).toMatchSnapshot();
    });
    
    it('renders footer content', () => {
        const wrapper = featurePanel(),
            footer = (wrapper.instance() as FeaturePanel).onRenderFooterContent();
        expect(footer.props.items).not.toHaveLength(0);
    });
});