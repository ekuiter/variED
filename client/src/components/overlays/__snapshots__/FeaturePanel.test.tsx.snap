// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeaturePanel renders information for a feature with a description 1`] = `
<StyledPanelBase
  headerText={
    <span>
      Feature
      : 
      <strong>
        FeatureIDE
      </strong>
    </span>
  }
  isLightDismiss={true}
  isOpen={true}
  onDismissed={[Function]}
  onRenderFooterContent={[Function]}
  type={1}
>
  <p>
    A sample description
  </p>
</StyledPanelBase>
`;

exports[`FeaturePanel renders information for a feature without a description 1`] = `
<StyledPanelBase
  headerText={
    <span>
      Feature
      : 
      <strong>
        FeatureHouse
      </strong>
    </span>
  }
  isLightDismiss={true}
  isOpen={true}
  onDismissed={[Function]}
  onRenderFooterContent={[Function]}
  type={1}
>
  <p>
    <em>
      No description set.
    </em>
  </p>
</StyledPanelBase>
`;
